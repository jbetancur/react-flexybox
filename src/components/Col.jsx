import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { media, genWidth } from '../utils/media';

const AgnosticComponent = ({
  alignSelf,
  flex,
  xs,
  sm,
  md,
  lg,
  gutter,
  columnDivisions,
  order,
  minWidth,
  debug,
  component,
  ...rest
}) => {
  const Tag = component;

  return <Tag {...rest} />;
};

const ColStyle = styled(AgnosticComponent)`
  box-sizing: border-box;
  margin: ${props => props.gutter}px;
  max-height: 100%;
  order: ${props => props.order};
  align-self: ${props => props.alignSelf};
  ${props => props.debug && 'border: 1px solid green'};
  ${props => props.flex && media.flex`
    --width: ${genWidth(props.flex, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};

  /* if flex is just true then set flex to auto */
  ${props => !Number.isInteger(props.flex) && props.flex && 'flex: 1'};
  ${props => Number.isInteger(props.xs) && media.xs`
    --width: ${genWidth(props.xs, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};
  ${props => Number.isInteger(props.sm) && media.sm`
    --width: ${genWidth(props.sm, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};
  ${props => Number.isInteger(props.md) && media.md`
    --width: ${genWidth(props.md, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};
  ${props => Number.isInteger(props.lg) && media.lg`
    --width: ${genWidth(props.lg, props.columnDivisions, props.gutter)};
    flex: 0 1 var(--width);
    min-width: var(--width);
  `};

  /* Allows min-width to be overridden */
  ${props => props.minWidth && `min-width: ${props.minWidth}px`};
`;

const Col = ({
  alignSelf,
  flex,
  xs,
  sm,
  md,
  lg,
  gutter,
  columnDivisions,
  order,
  className,
  style,
  minWidth,
  debug,
  component,
  children,
  ...rest,
}) => (
  <ColStyle
    alignSelf={alignSelf}
    flex={flex}
    xs={xs}
    sm={sm}
    md={md}
    lg={lg}
    gutter={gutter}
    columnDivisions={columnDivisions}
    order={order}
    className={className}
    style={style}
    minWidth={minWidth}
    debug={debug}
    component={component}
    {...rest}
  >
    {children}
  </ColStyle>
);


Col.propTypes = {
  /**
   * react/html elements
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),

  /**
   * aligns the whole structure according to its value and has no effect when items are in a single line
   */
  alignSelf: PropTypes.oneOf(['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch']),

  /**
   * inherited from `Row`
   */
  gutter: PropTypes.number,

  /**
   * inherited from `Row`
   */
  columnDivisions: PropTypes.number,

  /**
   * The component to render the `Col` as
   */
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),

  /**
   * inherited from `Row`
   */
  debug: PropTypes.bool,

  /**
   * allows overriding the `Col` item order
   */
  order: PropTypes.number,

  /**
   * sets the min-width on a `Col` item - this overrides `Row` `minColWidths` for a specific `Col`
   */
  minWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * overide the `className` on the root element of Row
   */
  className: PropTypes.string,

  /**
   * override the `style` on the root element of Row
   */
  style: PropTypes.object,

  /**
   * sets the default flex-basis and min-width percentage of the flex-item
   */
  flex: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 0-599-px
   */
  xs: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 600-959px
   */
  sm: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 960-1280px
   */
  md: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * sets the default flex-basis and min-width percentage of the flex-item when the screen size is 1280px or greater
   */
  lg: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),
};

Col.defaultProps = {
  alignSelf: 'auto',
  children: [],
  order: 0,
  minWidth: false,
  className: '',
  style: {},
  component: 'div',
  flex: false,
  xs: false,
  sm: false,
  md: false,
  lg: false,
  gutter: 0,
  columnDivisions: 12,
  debug: false,
};

Col.displayName = 'Col';

export default Col;
