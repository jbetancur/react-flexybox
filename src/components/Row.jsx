import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const AgnosticComponent = ({
  wrap,
  direction,
  justifyContent,
  alignContent,
  alignItems,
  minColWidths,
  debug,
  height,
  padding,
  paddingTop,
  paddingLeft,
  paddingBottom,
  paddingRight,
  center,
  fill,
  component,
  gutter,
  columnDivisions,
  ...rest
}) => {
  const Tag = component;

  return <Tag {...rest} />;
};

const RowStyle = styled(AgnosticComponent)`
  box-sizing: border-box;
  display: flex;
  flex: 0 1 auto;
  width: 100%;
  flex-direction: ${props => props.direction};
  flex-wrap: ${props => props.wrap};
  justify-content: ${props => (props.center ? 'center' : props.justifyContent)};
  align-items: ${props => (props.center ? 'center' : props.alignItems)};
  align-content: ${props => props.alignContent};
  ${props => props.debug && 'border: 1px solid red'};
  ${props => props.paddingTop && `padding-top: ${props.paddingTop}`};
  ${props => props.paddingLeft && `padding-left: ${props.paddingLeft}`};
  ${props => props.paddingBottom && `padding-bottom: ${props.paddingBottom}`};
  ${props => props.paddingRight && `padding-right: ${props.paddingRight}`};
  ${props => props.padding && `padding: ${props.padding}`};
  ${props => props.height && `height: ${props.height}`};
  ${props => props.fill && 'height: 100%'};
`;

const Row = (props) => {
  const {
    gutter,
    columnDivisions,
    debug,
    minColWidths,
    wrap,
    direction,
    justifyContent,
    alignContent,
    alignItems,
    className,
    style,
    height,
    padding,
    paddingTop,
    paddingLeft,
    paddingBottom,
    paddingRight,
    center,
    fill,
    component,
    children,
    ...rest
  } = props;

  const renderChildren = () => (
    React.Children.map(children, (child) => {
      if (child && child.type.displayName === 'Col') {
        return React.cloneElement(child, {
          gutter,
          columnDivisions,
          debug,
          minWidth: minColWidths,
        });
      }

      return child;
    })
  );

  return (
    <RowStyle
      wrap={wrap}
      direction={direction}
      justifyContent={justifyContent}
      alignContent={alignContent}
      alignItems={alignItems}
      minColWidths={minColWidths}
      className={className}
      style={style}
      debug={debug}
      height={height}
      padding={padding}
      paddingTop={paddingTop}
      paddingLeft={paddingLeft}
      paddingBottom={paddingBottom}
      paddingRight={paddingRight}
      center={center}
      fill={fill}
      component={component}
      {...rest}
    >
      {renderChildren(props)}
    </RowStyle>
  );
};

Row.propTypes = {
  /**
   * children
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),

  /**
   * sets the margins for flex items
   */
  gutter: PropTypes.number,

  /**
   * controls the size of each flex division. 1/12 divisions yeilds ~8% section of screen size.
   * 1/24 would yield ~4%
   */
  columnDivisions: PropTypes.number,

  /**
   * sets the min-width for all `Col` flex items. This also forces wrapping when the screen size is adjusted
   * and the flex item reaches its min-width
   */
  minColWidths: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),

  /**
   * overide the `className` on the root element of Row
   */
  className: PropTypes.string,

  /**
   * override the `style` on the root element of Row
   */
  style: PropTypes.object,

  /**
   * set the wrapping type
   */
  wrap: PropTypes.oneOf(['nowrap', 'wrap', 'wrap-reverse']),

  /**
   * The component to render the `Row` as
   */
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),

  /**
   * set the `Row` direction
   */
  direction: PropTypes.oneOf(['row', 'row-reverse', 'column', 'column-reverse']),

  /**
   * jusify `Col` items on the min x axis (hotizontally)
   */
  justifyContent: PropTypes.oneOf(['flex-start', 'flex-end', 'center', 'space-between', 'space-around', 'space-evenly']),

  /**
   * aligns the whole structure according to its value and has no effect when items are in a single line
   */
  alignContent: PropTypes.oneOf(['flex-start', 'flex-end', 'center', 'baseline', 'stretch']),

  /**
   * aligns the items inside a flex container along the y axis just like justifyContent does along the x axis
   */
  alignItems: PropTypes.oneOf(['flex-start', 'flex-end', 'center', 'space-between', 'space-around', 'stretch', 'baseline']),

  /**
   * enables bounding boxes for `Row` and `Col` to help debug layout issues
   */
  debug: PropTypes.bool,

  /**
   * set the height of the `Row`
   */
  height: PropTypes.string,

  /**
   * set the padding of the `Row`
   */
  padding: PropTypes.string,

  /**
   * set the padding-top of the `Row`
   */
  paddingTop: PropTypes.string,

  /**
   * set the padding-left of the `Row`
   */
  paddingLeft: PropTypes.string,

  /**
   * set the padding-bottom of the `Row`
   */
  paddingBottom: PropTypes.string,

  /**
   * set the padding-right of the `Row`
   */
  paddingRight: PropTypes.string,

  /**
   * shortcut to center the `Col` items within a `Row`
   */
  center: PropTypes.bool,

  /**
   * shortcut to fill the `Row` to `height: 100%`
   */
  fill: PropTypes.bool,
};

Row.defaultProps = {
  children: [],
  gutter: 0,
  columnDivisions: 12,
  minColWidths: false,
  className: '',
  style: {},
  wrap: 'wrap',
  component: 'div',
  direction: 'row',
  justifyContent: 'flex-start',
  alignContent: 'stretch',
  alignItems: 'stretch',
  debug: false,
  height: '',
  padding: '',
  paddingTop: '',
  paddingLeft: '',
  paddingBottom: '',
  paddingRight: '',
  center: false,
  fill: false,
};

Row.displayName = 'Row';

export default Row;
