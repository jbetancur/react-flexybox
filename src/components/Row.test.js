import React from 'react';
import { render } from 'react-testing-library';
import 'jest-styled-components';
import Row from './Row';
import Col from './Col';

describe('Row', () => {
  test('it mounts <Row />', () => {
    const { container } = render(<Row />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts Row with props', () => {
    const { container } = render(
      <Row
        paddingTop="1em"
        paddingLeft="1em"
        paddingBottom="1em"
        paddingRight="1em"
        padding="1em"
        height="100px"
        className="rick-morty"
        style={{ color: 'purple' }}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts Row and passes with debug prop to Col', () => {
    const { container } = render(
      <Row
        debug
      >
        <Col />
      </Row>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts Row with shortcut props', () => {
    const { container } = render(
      <Row
        fill
        center
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts Row with shortcut prop fill and overrides height if set', () => {
    const { container } = render(
      <Row
        height="10px"
        fill
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts Row with shortcut prop padding and overrides padding<direction> if set', () => {
    const { container } = render(
      <Row
        paddingTop="1em"
        paddingLeft="1em"
        paddingBottom="1em"
        paddingRight="1em"
        padding="0 1px 2px 0"
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it mounts <Row component="span" />', () => {
    const { container } = render(
      <Row component="span" />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it overrides the default component as a component', () => {
    const Faux = () => <div />;
    const { container } = render(
      <Row component={Faux} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it passes props down to a child that is a <Col />', () => {
    const { container } = render(
      <Row>
        <Col />
      </Row>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it passes props down to a child that is NOT <Col />', () => {
    const { container } = render(
      <Row>
        <section>test</section>
      </Row>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
