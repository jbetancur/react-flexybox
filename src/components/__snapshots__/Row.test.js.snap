// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row it mounts <Row /> 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

<div
  class="c0"
/>
`;

exports[`Row it mounts <Row component="span" /> 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

<span
  class="c0"
/>
`;

exports[`Row it mounts Row and passes with debug prop to Col 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  border: 1px solid red;
}

.c1 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  border: 1px solid green;
}

<div
  class="c0"
>
  <div
    class="c1"
  />
</div>
`;

exports[`Row it mounts Row with props 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  padding-top: 1em;
  padding-left: 1em;
  padding-bottom: 1em;
  padding-right: 1em;
  padding: 1em;
  height: 100px;
}

<div
  class="rick-morty c0"
  style="color: purple;"
/>
`;

exports[`Row it mounts Row with shortcut prop fill and overrides height if set 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  height: 10px;
  height: 100%;
}

<div
  class="c0"
/>
`;

exports[`Row it mounts Row with shortcut prop padding and overrides padding<direction> if set 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  padding-top: 1em;
  padding-left: 1em;
  padding-bottom: 1em;
  padding-right: 1em;
  padding: 0 1px 2px 0;
}

<div
  class="c0"
/>
`;

exports[`Row it mounts Row with shortcut props 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  height: 100%;
}

<div
  class="c0"
/>
`;

exports[`Row it overrides the default component as a component 1`] = `<div />`;

exports[`Row it passes props down to a child that is NOT <Col /> 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

<div
  class="c0"
>
  <section>
    test
  </section>
</div>
`;

exports[`Row it passes props down to a child that is a <Col /> 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

.c1 {
  box-sizing: border-box;
  margin: 0px;
  max-height: 100%;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

<div
  class="c0"
>
  <div
    class="c1"
  />
</div>
`;
